pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,777 @@
+	preEn := obj.(*v1alpha1.ExternalNode)
+	if reflect.DeepEqual(preEn.Spec.Interfaces[0], curEn.Spec.Interfaces[0]) {
+		klog.InfoS("Skip processing ExternalNode update as no changes for Interface[0]", "ExternalNode", klog.KObj(curEn))
+		return nil
+	}
+	preEEName, err := externalnode.GenExternalEntityName(preEn)
+	if err != nil {
+		return err
+	}
+	preIfName, preIPs, err := getHostInterfaceName(preEn.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	curEEName, err := externalnode.GenExternalEntityName(curEn)
+	if err != nil {
+		return err
+	}
+	curIfName, curIPs, err := getHostInterfaceName(curEn.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	if preIfName != curIfName {
+		klog.InfoS("Found interface name is changed", "preName", preIfName, "curName", curIfName)
+		if err = c.addInterface(curIfName, curEn.Namespace, curEEName, curIPs); err != nil {
+			return err
+		}
+		if err = c.deleteInterface(preIfName); err != nil {
+			return err
+		}
+	} else if (!reflect.DeepEqual(preIPs, curIPs)) || (preEEName != curEEName) {
+		klog.InfoS("Found interface configuration is changed", "preIPs", preIPs, "preEEName", preEEName,
+			"curIPs", curIPs, "curEEName", curEEName)
+		if err = c.addInterface(curIfName, curEn.Namespace, curEEName, curIPs); err != nil {
+			return err
+		}
+	}
+	c.syncedExternalNodes.Add(curEn)
+	// Notify the ExternalNode update event to NetworkPolicyController.
+	c.nodeUpdateNotifier.Notify(v1beta2.ExternalEntityReference{
+		Name:      curEEName,
+		Namespace: curEn.Namespace,
+	})
+	return nil
+}
+
+func (c *ExternalNodeController) deleteExternalNode(key string) error {
+	klog.InfoS("Deleting ExternalNode", "key", key)
+	obj, exists, _ := c.syncedExternalNodes.GetByKey(key)
+	if !exists {
+		klog.InfoS("Skipping ExternalNode deletion as it hasn't been synced", "ExternalEntityKey", key)
+		return nil
+	}
+	en := obj.(*v1alpha1.ExternalNode)
+	ifName, _, err := getHostInterfaceName(en.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	if err := c.deleteInterface(ifName); err != nil {
+		return err
+	}
+	c.syncedExternalNodes.Delete(en)
+	return nil
+}
+
+func (c *ExternalNodeController) deleteInterface(ifName string) error {
+	hostIface, ifaceExists := c.ifaceStore.GetInterfaceByName(ifName)
+	if !ifaceExists {
+		klog.InfoS("Skipping deleting host interface since it doesn't exist ", "ifName", ifName)
+		return nil
+	}
+	klog.InfoS("Deleting interface", "ifName", ifName)
+	if err := c.removeOVSPortsAndFlows(hostIface); err != nil {
+		return err
+	}
+	c.ifaceStore.DeleteInterface(hostIface)
+	return nil
+}
+
+func (c *ExternalNodeController) createOVSPortsAndFlows(uplinkName, hostIfName, eeNamespace string, eeName string, ips []string) (*interfacestore.InterfaceConfig, error) {
+	adapterConfig, err := util.GetUplinkConfig(hostIfName)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get the configuration on the host interface", "hostInterface", hostIfName)
```

Comment:
Paste my previous comment to another PR:

> For all the error logs, could you check we do need to log them, or we can append the error message to the returned error, and log in the caller? It is ok in some cases you like to log twice with more information where the error first happens, but we need to avoid many duplicated error logs, and esp. if you just return the error without prepending extra context from the callee, it may generate confusing log in the caller (as the caller usually logs the returned error).
> 
> For example, assume func A calls B calls C, and C gets a "not-found" error. If all A, B, C log the error and directly return the error to the caller, we will get: "A: not found", "B: not found", "C: sth is not found". First, it is unnecessary to have the same error logged three times; second, just "not-found" without further information could be confusing in the logs in A and B. So, we usually prepend the error context to the returned error and log in the caller, e.g. in C - return fmt.Errorf("failed to do sth: %v", err)